Enemy System and Player Interaction

MurilloYonamine opened the pull request:
Add basic player-enemy interaction system with movement, damage, and knockback

This PR introduces the foundation of the interaction between the player and enemies, including movement, damage logic, and knockback effects.

Main changes:

- Created an IDamageable interface, allowing all entities that can take damage to implement this logic.
- Implemented knockback logic when taking damage.
- Started the enemy logic, including:
- Simple movement towards the player.
- Damage system and knockback.
- Added a Cinemachine Camera in the scene to follow the player’s position.


![gameplay-preview](https://github.com/user-attachments/assets/594c3341-1859-4b6e-a396-d1618b5c0668)

MurilloYonamine committed:
Added New Input System to the project

- I imported the new input system.
- Configured my inputs.

MurilloYonamine committed:
Added a scene and created a Input Handler Script

I created a script that will responsible to handle all player inputs.

Also I made a scene to start building this project.

MurilloYonamine committed:
Add enemy and player interaction with damage handling and knockback effects

- Created an interface called "IDamageable", all entities that take damage will have this implementation.
- Also created knockback logic.
- And started the enemy logic.

MurilloYonamine committed:
Add Enemy movement towards the player

- Just add a really simple movement towards the player.
- Also add a Cinemachine Camera that will follow player position.

Harsche commented on Assets/_Project/Prototype Pattern/Scripts/IDamageable.cs:
```diff
@@ -0,0 +1,44 @@
+        /// <summary>
+        /// Gets or sets the current health value of the object.
+        /// </summary>
+        public float Health { get; set; }
```
Use public setters only when necessary

MurilloYonamine replied:
```
 public float Health
        {
            get => _health;
            set
            {
                _health = value;
                if (_health <= 0)
                {
                    Targetable = false;
                    OnDeath();
                }
            }
        }
```

I used them on my PlayerController and probably I am gonna use them on enemy's

Harsche commented on Assets/_Project/Prototype Pattern/Scripts/Player/PlayerController.cs:
```diff
@@ -0,0 +1,80 @@
+            }
+        }
+
+        private void Start()
```
Always keep initialization logic in Awake.

```suggestion
        private void Awake()
        {
            _physicsCollider = GetComponent<Collider2D>();
            _rigidBody2D = GetComponent<Rigidbody2D>();
            _inputHandler = GetComponent<InputHandler>();
        }
```

MurilloYonamine replied:
I have a question!

I avoid using anything in Awake, because I never know when I am gonna have a Singleton and I don't want to have problem with initialization.

So should I have in my mind to always initialize "GetComponent logic" in Awake? Are there another that I need to keep in my mind?

To be honest, I really don't know what should be in Start or what should be in Awake.

Harsche commented on Assets/_Project/Prototype Pattern/Scripts/Player/PlayerController.cs:
```diff
@@ -0,0 +1,80 @@
+            if (Targetable && !Invencible)
+            {
+                Health -= damage;
+                _inputHandler.ApplyKnockback(knockback, 0.1f);
```
You probably want to make the duration a serialized field for design purposes.

Harsche commented on Assets/_Project/Prototype Pattern/Scripts/Player/InputHandler.cs:
```diff
@@ -0,0 +1,93 @@
+        [SerializeField] private float _speed = 20f;
+        [SerializeField] private bool _isInKnockback = false;
+
+        private void Start()
```
```suggestion
        private void Awake()
```

Harsche commented on Assets/_Project/Prototype Pattern/Scripts/Player/InputHandler.cs:
```diff
@@ -0,0 +1,93 @@
+
+        private Rigidbody2D _rigidBody2D;
+
+        [SerializeField] private Vector2 direction;
```
```suggestion
        [SerializeField] private Vector2 _direction;
```

Harsche commented on Assets/_Project/Prototype Pattern/Scripts/Player/InputHandler.cs:
```diff
@@ -0,0 +1,93 @@
+        {
+            if (!_player.Invencible)
+            {
+                _rigidBody2D.velocity = new Vector2(direction.x * _speed, direction.y * _speed);
```
Why not multiply the vector directly?

```suggestion
                _rigidBody2D.velocity = _direction * _speed;
```

Harsche commented on Assets/_Project/Prototype Pattern/Scripts/Player/InputHandler.cs:
```diff
@@ -0,0 +1,93 @@
+
+        private void OnEnable()
+        {
+            _playerControls = new PlayerControls();
```
Would it be better to instantiate it in Awake and just enable/disable it later?

MurilloYonamine replied:
Isn't it the same thing? I thought it was.

I will fix this later!

Harsche commented on Assets/_Project/Prototype Pattern/Scripts/Enemy/EnemyController.cs:
```diff
@@ -0,0 +1,34 @@
+    {
+        if (_player != null)
+        {
+            transform.position = Vector2.MoveTowards(transform.position, _player.gameObject.transform.position, _speed * Time.deltaTime);
```
More readable


```suggestion
            Vector2 playerPosition = _player.transform.position;
            transform.position = Vector2.MoveTowards(transform.position, playerPosition, _speed * Time.deltaTime);
```

Harsche reviewed (changes requested):
Review requested changes. You can also reply anything you don't agree with or have a better suggestion.

MurilloYonamine committed:
All the requestes changes have been resolved.

Requests like:

- Invincible misspelling.
- Add "_" in the direction variable.
- Changed Start to Awake because of "GetComponent" logic.
- Changed Debug.log messages that was in Portuguese to English.
- Now the duration of the knockback animation has its own variable for design purposes.
- Also changed a number in EditorSettings file to avoid conflicts.

MurilloYonamine committed:
Removed setters from IDamageable interface.

PlayerController Health and Targetable setter are now private.

Harsche commented on Assets/_Project/Prototype Pattern/Scripts/Player/PlayerController.cs:
```diff
@@ -0,0 +1,80 @@
+        [SerializeField] private bool _invincible = false;
+        [SerializeField] private Collider2D _physicsCollider;
+
+        [SerializeField] private Rigidbody2D _rigidBody2D;
```
Any reason to keep them serialized when they're also being assigned in Awake?

MurilloYonamine replied:
no reason. I will remove them.

Harsche reviewed (approved):
Great job!
